Mathlib/Topology/Algebra/Subgroup.lean

```diff
@@ -364,3 +365,219 @@ theorem isOpen_of_isOpen_subideal {U I : Ideal R} (h : U ≤ I) (hU : IsOpen (U
+    (fun (H : OpenNormalSubgroup G) ↦ ↑H.toOpenSubgroup) (fun _ _ => rfl)
+
+@[to_additive]
+instance [ContinuousMul G] : Lattice (OpenNormalSubgroup G) :=
+  { instSemilatticeInfOpenNormalSubgroup,
+    instSemilatticeSupOpenNormalSubgroup with
+    toPartialOrder := instPartialOrderOpenNormalSubgroup}
+
+end OpenNormalSubgroup
+
+end
+
+/-!
+# Closed subgroups of a topological groups
+
+This section builds the SemilatticeInf `ClosedSubgroup G` of closed subgroups in a
+topological group `G`, and its additive version `ClosedAddSubgroup`.
+
+# Main definitions and results
+
+* `normalCore_isClosed` : The `normalCore` of a closed subgroup is closed.
+
+*  `finiteindex_closedSubgroup_isOpen` : A closed subgroup with finite index is open.
+
+-/
+
+section
+
+universe u v
+
+/-- The type of open subgroups of a topological group. -/
+@[ext]
+structure ClosedSubgroup (G : Type u) [Group G] [TopologicalSpace G] extends Subgroup G where
+  isClosed' : IsClosed carrier
+
+/-- The type of open subgroups of a topological group. -/
+@[ext]
+structure ClosedAddSubgroup (G : Type u) [AddGroup G] [TopologicalSpace G] extends
+    AddSubgroup G where
+  isClosed' : IsClosed carrier
+
+attribute [to_additive] ClosedSubgroup
+
+attribute [coe] ClosedSubgroup.toSubgroup ClosedAddSubgroup.toAddSubgroup
+
+namespace ClosedSubgroup
+
+variable (G : Type u) [Group G] [TopologicalSpace G]
+
+variable {G} in
+@[to_additive]
+theorem toSubgroup_injective : Function.Injective
+    (ClosedSubgroup.toSubgroup : ClosedSubgroup G → Subgroup G) :=
+  fun A B h => by
+  ext
+  rw [h]
+
+@[to_additive]
+instance : SetLike (ClosedSubgroup G) G where
+  coe U := U.1
+  coe_injective' _ _ h := toSubgroup_injective <| SetLike.ext' h
+
+@[to_additive]
+instance : SubgroupClass (ClosedSubgroup G) G where
+  mul_mem := Subsemigroup.mul_mem' _
+  one_mem U := U.one_mem'
+  inv_mem := Subgroup.inv_mem' _
+
+@[to_additive]
+instance : Coe (ClosedSubgroup G) (Subgroup G) where
+  coe := toSubgroup
+
+@[to_additive]
+instance : PartialOrder (ClosedSubgroup G) := inferInstance
+
+@[to_additive]
+instance instInfClosedSubgroup : Inf (ClosedSubgroup G) :=
+  ⟨fun U V => ⟨U ⊓ V, U.isClosed'.inter V.isClosed'⟩⟩
+
+@[to_additive]
+instance instSemilatticeInfClosedSubgroup : SemilatticeInf (ClosedSubgroup G) :=
+  SetLike.coe_injective.semilatticeInf ((↑) : ClosedSubgroup G → Set G) fun _ _ => rfl
+
+end ClosedSubgroup
+
+open scoped Pointwise
+
+namespace TopologicalGroup
+
+variable {G : Type u} [Group G] [TopologicalSpace G] [ContinuousMul G]
+
+lemma normalCore_isClosed (H : Subgroup G) (h : IsClosed (H : Set G)) :
+    IsClosed (H.normalCore : Set G) := by
+  have : H.normalCore = ⨅ (g : ConjAct G), g • H := by
+    ext g
+    simp [Subgroup.normalCore, Subgroup.mem_iInf, Subgroup.mem_pointwise_smul_iff_inv_smul_mem]
+    constructor
+    <;> intro h x
+    <;> specialize h (x⁻¹)
+    · exact h
+    · simp only [ConjAct.toConjAct_inv, inv_inv] at h
+      exact h
```

Comment:
This is what `simpa` is for (you can surely do this in one line instead of two).

Reply:
fixed